poem/src/server.rs

```diff
@@ -135,20 +138,25 @@ where
                         let ep = ep.clone();
                         let alive_connections = alive_connections.clone();
                         let notify = notify.clone();
-                        let timeout_notify = timeout_notify.clone();
+                        let timeout_token = timeout_token.clone();
+                        let conn_shutdown_token = conn_shutdown_token.clone();
 
                         tokio::spawn(async move {
+                            let serve_connection = serve_connection(socket, local_addr, remote_addr, scheme, ep, conn_shutdown_token);
+
                             if timeout.is_some() {
                                 tokio::select! {
-                                    _ = serve_connection(socket, local_addr, remote_addr, scheme, ep) => {}
-                                    _ = timeout_notify.notified() => {}
+                                    _ = serve_connection => {}
+                                    _ = timeout_token.cancelled() => {}
                                 }
                             } else {
-                                serve_connection(socket, local_addr, remote_addr, scheme, ep).await;
+                               serve_connection.await;
                             }
 
                             if alive_connections.fetch_sub(1, Ordering::Acquire) == 1 {
-                                notify.notify_one();
+                                // we have to notify only if there is a registered waiter on shutdown.
+                                // `notify_one` will make `notified` completed immidiately as soon as we don't have at least 1 connection
+                                notify.notify_waiters();
```

Comment:
It was broken logic actually.

Once you have handled some number of connections - it will `notify_one` in advance. So gracefult shutdown didn't work as expected and just passed `notified()` immediately without waiting for notification